File: algorithms/179.cpp
#include <iostream>
#include <sstream>
#include <vector>
#include <algorithm>
using namespace std;

class Solution
{
private:
    static bool cmp(const string &a, const string &b)
    {
        if ( a.size() != b.size() )
            return cmp(a+b, b+a);
        else
            return a < b;
    }
public:
    string largestNumber(vector<int>& nums)
    {
        vector<string> strnums(nums.size());
        for ( unsigned int i = 0; i < nums.size(); ++i )
        {
            stringstream ss;
            ss << nums[i];
            strnums[i] = ss.str();
        }
        sort(strnums.begin(), strnums.end(), cmp);
        stringstream result;
        for ( int i = nums.size()-1; i >= 0; --i )
            result << strnums[i];
        if ( result.str()[0] == '0' )
            return "0";
        else
            return result.str();
    }
};

int main()
{
    return 0;
}

File: algorithms/python/179.py
from functools import cmp_to_key

class Solution:
    def largestNumber(self, nums):
        """
        :type nums: List[int]
        :rtype: str
        """
        nums = sorted(map(str, nums), key=cmp_to_key(self.cmp))
        return str(int(''.join(nums)))
        
    def cmp(self, x, y):
        a, b = int(x+y), int(y+x)
        if a < b:
            return 1
        elif a > b:
            return -1
        else:
            return 0
